### DISCUSSION TITLE
Making Functions working with `Lazy<T>` trimmer friendly

### DISCUSSION THREAD

**Discussion by tautropfli** [dotnet/runtime#69486] [Q&A] [answered]
I'm trying to make these two functions trimmer friendly:

```cs
public static class Lazy
{
	public static Lazy<TItem> Return<TItem>(TItem value) => new(value);

	public static Lazy<TResult> Select<T, TResult>(this Lazy<T> lazy, Func<T, TResult> selector)
		=> new(() => selector(lazy.Value));
}
```

`dotnet publish` with trimming enabled gives these warnings:

```
Lazy.cs(x,y): Trim analysis error IL2091: Lazy.Return<TItem>(TItem): 'T' generic argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicParameterlessConstructor' in 'System.Lazy<T>'. The generic parameter 'TItem' of 'Lazy.Return<TItem>(TItem)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
Lazy.cs(x,y): Trim analysis error IL2091: Lazy.Select<T,TResult>(Lazy<T>,Func<T,TResult>): 'T' generic argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicParameterlessConstructor' in 'System.Lazy<T>'. The generic parameter 'TResult' of 'Lazy.Select<T,TResult>(Lazy<T>,Func<T,TResult>)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
```

Is it OK to use `[UnconditionalSuppressMessage]` to suppress them?

I think it should be OK because:
1. I'm always creating `Lazy<T>` with either a `value` or `valueFactory`.
2. My code doesn't use the public parameterless constructor itself.

**Comment by vitek-karas** [comment, marked as answer]
I think in this case it's probably safe, but I would still suggest considering annotating the generic arguments of your methods to avoid the suppression. Suppressions are very tricky to get right and even if they are right at the time you add them, they may not be forever. If somebody changes the code to actually do call the public parameterless .ctor there will still be no warning (because of the suppression) and suddenly things will break.

I vaguely remember us discussing that some .ctors of `Lazy` don't need the annotation (as you mention), but currently there's no good way to express this fact in the language/attributes.

**Reply by tautropfli** [reply]
> Suppressions are very tricky to get right and even if they are right at the time you add them, they may not be forever.

Ah, that's a good point. Thank you!
